Support 8-bit numbers in RGB

mkko opened the pull request:
Fix for #7. The number formatting is now checked before parsing. Alpha didn't need this, it's always a floating point value, but it doesn't hurt to check that too.

mkko committed:
Support 8-bit numbers in RGB

samrayner reviewed (changes requested):
Thanks for this @mkko. Your changes drop support for non-string values in the JSON that were produced by Xcode 9 and 9.1. Would you mind amending it to maintain support?

Also, please could you add screens to the example app that use a colour set to 8-bit mode and a colour set to integer mode? The example app acts as a test suite of sorts (tap the screen to cycle through colours).

mkko commented:
Can you be more specific? Which values would be non-string values? I tried using Xcode 9.0 and it writes the values as strings, the same way as 9.4.1.

And sure, I'll add specific items for the 8-bit support to the example.

mkko committed:
Add 8-bit colors to the example

samrayner reviewed (changes requested):
Some of the colour assets in the example apps were generated by Xcode with non-quoted values (see the contents.json of "universal").

Testing your code after a Clean (to make sure the Pod changes are picked up) the first few colours in the example display as black because the decoding fails.

I thought that the generated JSON having string rather float values was a change that came in Xcode 9.2 but I might be wrong.

mkko commented:
I still can't reproduce the numeric values. Using Xcode 9.0 produces strings the same way a Xcode 9.4.1. Could it be that they changed it somewhere in between and changed it back? Doesn't really sound like Apple though. Do you have a version of Xcode to reproduce this with?

Anyways I don't think there's harm adding support for numeric JSON values as well. However, I'm not sure if there's a way to differentiate between floating point and integer values.

I would start by parsing a string and fallback to a numeric. What do you think?

mkko committed:
Use numeric parsing as a fallback

mkko commented:
I added the numeric parsing to make the examples work. However, I couldn't produce any numeric values with Xcodes 9.0 or 9.4.1. I didn't try any versions in between those two.

There's one caveat when parsing numeric values though; There's no way of telling which base the value is using (0-1 or 0-255), since from the example it looks like it is not enough to check whether the value is integer or not. I added a logic where any values greater than 1 are considered to be values 8-bit values. This logic breaks if something should be 1/255 and is represented as an integer, but I would consider this to be an edge case of an edge case.

samrayner commented:
Awesome, thanks @mkko!
